### PR TITLE
Update `InMemoryState` and `SqliteState` for the Ping protocol.

### PR DESCRIPTION
**Pull request by panh99**

### CODE REVIEW

File: src/py/flwr/server/superlink/state/in_memory_state.py [hunk 1]
```diff
@@ -210,17 +215,32 @@ def get_nodes(self, run_id: int) -> Set[int]:
...
-        return self.node_ids
+        with self.lock:
+            if run_id not in self.run_ids:
+                return set()
+            current_time = time.time()
+            return {
+                node_id
+                for node_id, online_until in self.node_ids.items()
+                if online_until > current_time
+            }
```
Comment (jafermarq) [inline src/py/flwr/server/superlink/state/in_memory_state.py:226]: Wouldn't it be better to do this like:
```
            return {
                node_id
                for node_id, last_online in self.node_ids.items()
                if last_online > current_time - 30 # or other ping interval
            }
```

File: src/py/flwr/server/superlink/state/in_memory_state.py [hunk 2]
```diff
@@ -210,17 +215,32 @@ def get_nodes(self, run_id: int) -> Set[int]:
...
+                self.run_ids.add(run_id)
+                return run_id
         log(ERROR, "Unexpected run creation failure.")
         return 0
+
+    def acknowledge_ping(self, node_id: int, ping_interval: float) -> bool:
+        """Acknowledge a ping received from a node, serving as a heartbeat."""
+        with self.lock:
+            if node_id in self.node_ids:
+                self.node_ids[node_id] = time.time() + ping_interval
```
Comment (jafermarq) [inline src/py/flwr/server/superlink/state/in_memory_state.py:244]: why add the `ping_interval` to the timestamp here? wouldn't it be better to use the interval just when doing `get_nodes()

File: src/py/flwr/server/superlink/state/in_memory_state.py [hunk 3]
```diff
@@ -190,17 +191,21 @@ def create_node(self) -> int:
         # Sample a random int64 as node_id
         node_id: int = int.from_bytes(os.urandom(8), "little", signed=True)
 
-        if node_id not in self.node_ids:
-            self.node_ids.add(node_id)
-            return node_id
+        with self.lock:
+            if node_id not in self.node_ids:
+                # Default ping interval is 30s
+                # TODO: change 1e9 to 30s  # pylint: disable=W0511
```
[the remark this answers is not in the document]
Comment (danieljanes) [reply src/py/flwr/server/superlink/state/in_memory_state.py:198]: Should we set a default here, or add `ping_interval` to `CreateNodeRequest`?

File: src/py/flwr/server/superlink/state/sqlite_state.py [hunk 4]
```diff
@@ -468,9 +476,14 @@ def create_node(self) -> int:
...
 
-        query = "INSERT INTO node VALUES(:node_id);"
+        query = (
+            "INSERT INTO node (node_id, online_until, ping_interval) VALUES (?, ?, ?)"
+        )
+
         try:
-            self.query(query, {"node_id": node_id})
+            # Default ping interval is 30s
+            # TODO: change 1e9 to 30s  # pylint: disable=W0511
```
Comment (danieljanes) [inline src/py/flwr/server/superlink/state/sqlite_state.py:485]: Can we already implement this?

File: src/py/flwr/server/superlink/state/sqlite_state.py [hunk 5]
```diff
@@ -515,6 +528,17 @@ def create_run(self) -> int:
         log(ERROR, "Unexpected run creation failure.")
         return 0
 
+    def acknowledge_ping(self, node_id: int, ping_interval: float) -> bool:
+        """Acknowledge a ping received from a node, serving as a heartbeat."""
+        # Update the `online_until` field for the given `node_id`
```
Comment (danieljanes) [inline src/py/flwr/server/superlink/state/sqlite_state.py:533]:
```suggestion
        # Update `online_until` and `ping_interval` for the given `node_id`
```

**Review by danieljanes — APPROVED** [review]

**Pull request merged**